mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java

```diff
@@ -0,0 +1,342 @@
+    assertThat(HierarchyScanner.resolveTypeHierarchy(c.classElement, context).toString())
+        .isEqualTo("[c, d, x, w, y, z, a]");
+  }
+
+  private MockClass c(String name, MockClass parent, MockInterface... interfaces) {
+    return new MockClass(name, null, null, parent, interfaces);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void should_build_property_hierarchy_with_strategy_defined_on_interface_include_highest() {
+    /*
+     * given the following hierarchy
+     *
+     *                  a
+     *                 /
+     *        a r     z
+     *       / /     /
+     *      b       y
+     *       \     /
+     *        d  @x w
+     *         \ / /
+     *          c
+     */
+    // with a HierarchyScanStrategy annotation on "x" that dictates that "y" is the highest class,
+    // but to include it.
+    HierarchyScanStrategy yHighest = Mockito.mock(HierarchyScanStrategy.class);
+    Mockito.when(yHighest.includeHighestAncestor()).thenReturn(true);
+    Mockito.when(yHighest.scanAncestors()).thenReturn(true);
+    MockInterface a = i("a");
+    MockInterface r = i("r");
+    MockInterface z = i("z", a);
+    MockInterface y = i("y", z);
+    MockInterface x = i("x", yHighest, y, y);
+    // when checking to see if we're at the 'highest' class (this.getClass()) return true
+    // for y.
+    Mockito.when(context.getClassUtils().isSame(y.classElement, this.getClass())).thenReturn(true);
+    MockInterface w = i("w");
+
+    MockClass b = c("b", null, a, r);
+    MockClass d = c("d", b);
+    MockClass c = c("c", d, x, w);
+
+    // y's parent interface z should be skipped, however a will also be encountered
+    // as it is still in the hierarchy of c -> d -> b -> a
+    assertThat(HierarchyScanner.resolveTypeHierarchy(c.classElement, context).toString())
+        .isEqualTo("[c, d, x, w, b, y, a, r]");
+  }
+
+  private MockClass c(
+      String name,
+      HierarchyScanStrategy strategy,
+      MockElement highestAncestor,
+      MockClass parent,
+      MockInterface... interfaces) {
+    return new MockClass(name, strategy, highestAncestor, parent, interfaces);
+  }
+
+  private MockInterface i(String name, MockInterface... interfaces) {
+    return new MockInterface(name, null, null, interfaces);
+  }
+
+  private MockInterface i(
+      String name,
+      HierarchyScanStrategy strategy,
+      MockElement highestAncestor,
+      MockInterface... interfaces) {
+    return new MockInterface(name, strategy, highestAncestor, interfaces);
+  }
+
+  private TypeMirror root() {
+    TypeMirror noneMirror = Mockito.mock(TypeMirror.class);
+    Mockito.when(noneMirror.getKind()).thenReturn(TypeKind.NONE);
+
+    return noneMirror;
+  }
+
+  class MockElement {
+    final String name;
+    final TypeMirror mirror;
+    final TypeElement classElement;
+    final List<MockElement> interfaces;
+    final MockClass parent;
+    final Name qfName;
+
+    @SuppressWarnings("unchecked")
+    MockElement(
```

Comment:
The amount of mocking needed ended up getting out of control, perhaps we could consider removing this test now given we have decent integration test coverage.